Add routines to prevent adding ICs/time derivatives twice in Physics

GiudGiud opened the pull request:
refs #30533

we can imagine the same thing for time integrators / ...

I added a bunch of utilities

I might split off some commits into separate PRs here

GiudGiud committed:
Add more constness to material system

GiudGiud committed:
Add a routine to decide whether to create variables based on whether the variable already
exits and has the adequate block restriction
refs #30533

GiudGiud committed:
Add sorting by variables in the moose object warehouses:
will be used for ICs and time kernels, at least
refs #30533

GiudGiud committed:
Add a routine to determine whether to add an IC or not for a Physics
refs #30533

GiudGiud committed:
Add utils and routines to work with sets of subdomains

GiudGiud committed:
Add support for FV variables in 'shouldIC' routine for Physics
refs #30533

GiudGiud committed:
Add capability for warehouses to see if objects exists for sets of blocks for a given variable
refs #30533

GiudGiud committed:
Add a routine to determine whether to add a time derivative or not for a Physics
refs #30533

GiudGiud committed:
Deploy shouldCreateIC and shouldCreateTimeDerivative to all physics
refs #30533

GiudGiud commented on framework/include/interfaces/BlockRestrictable.h:
```diff
@@ -202,6 +209,11 @@ class BlockRestrictable
+  /**
+   * Adds blocks to the block restriction of the object
+   */
+  void addBlocks(const std::vector<SubdomainName> & new_blocks);
```
this is potentially controversial so it's in its own commit. 
happy to separate it if needed

GiudGiud replied:
This is a pretty simple design. An alternative could be to make a "MutableBlockRestrictable" interface which inherits from BlockRestrictable and is used only for objects we know benefit from a modified BR (e.g. mostly variables! but maybe the NS Rhie Chow user objects too)

GiudGiud replied:
ok it also does not work as the variable block restriction is given to the system at creation time. 
Maybe that can be changed. But for now I m removing this commit in the interest of getting this merged

lindsayad replied:
yea definitely would prefer to remove this until it's determined it's needed

GiudGiud replied:
ok done!

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2876539/), step Clang format on b5ca330 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30534/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30534/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8fad5bbd8699370cefab6bdfef6987ee74bf96b4`

GiudGiud committed:
Simplify shouldCreateVariable
Add a routine to report on unused parameters
refs #30533

GiudGiud committed:
Deploy shouldCreateVariables to all Physics, including calls to
reportPotentiallyMissedParams to warn user of missed parameters when using another parameter
refs #30533

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2877084/), step Clang format on ecb3c6a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30534/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30534/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d044f2fa58253ddc3442f38818684a97949868eb`

GiudGiud committed:
Add threading to new warehouse sorting

GiudGiud commented on framework/src/physics/PhysicsBase.C:
```diff
@@ -477,3 +480,206 @@ PhysicsBase::addPetscPairsToPetscOptions(
+    has_all_blocks = _problem->getFVInitialConditionWarehouse().hasObjectsForVariableAndBlocks(
+        var_name, blocks_set, blocks_covered, 0);
+  else
+    has_all_blocks = _problem->getInitialConditionWarehouse().hasObjectsForVariableAndBlocks(
```
so as you cant see here and below, using the dedicated warehouse makes the code a lot smaller and readable than the query

the problem is that it adds to the dedicated warehouses, instead of developing theWarehouse() with new queries, which maybe we are not trying to do?

GiudGiud replied:
would love some guidance on this

GiudGiud committed:
Add more logic to handle covered blocks

GiudGiud committed:
Add missing prereq

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2878052/), step Docs: sync website on 129f5f3 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30534/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2879749/), step Generate coverage on e6ec9ff wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>831b92</th>
      <th colspan=4>#30534 e6ec9f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.47%</td>
      <td>85.39%</td>
      <td>-0.08%</td>
      <td>32.06%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>112588</td>
      <td>112634</td>
      <td>+46</td>
      <td>67</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19134</td>
      <td>19269</td>
      <td>+135</td>
      <td>142</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30534/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30534/coverage/framework)


## Modules coverage

### Heat transfer

<table>
  <thead>
    <tr>
      <th></th>
      <th>831b92</th>
      <th colspan=4>#30534 e6ec9f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.48%</td>
      <td>89.50%</td>
      <td>+0.02%</td>
      <td>94.34%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4499</td>
      <td>4536</td>
      <td>+37</td>
      <td>50</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>529</td>
      <td>532</td>
      <td>+3</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30534/coverage_diff/heat_transfer)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30534/coverage/heat_transfer)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>831b92</th>
      <th colspan=4>#30534 e6ec9f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>86.46%</td>
      <td>86.50%</td>
      <td>+0.04%</td>
      <td>86.81%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>18420</td>
      <td>18457</td>
      <td>+37</td>
      <td>79</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2885</td>
      <td>2881</td>
      <td>-4</td>
      <td>12</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30534/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30534/coverage/navier_stokes)

### Scalar transport

<table>
  <thead>
    <tr>
      <th></th>
      <th>831b92</th>
      <th colspan=4>#30534 e6ec9f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>92.01%</td>
      <td>92.03%</td>
      <td>+0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>380</td>
      <td>381</td>
      <td>+1</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>33</td>
      <td>33</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30534/coverage_diff/scalar_transport)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30534/coverage/scalar_transport)

### Stochastic tools

<table>
  <thead>
    <tr>
      <th></th>
      <th>831b92</th>
      <th colspan=4>#30534 e6ec9f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.83%</td>
      <td>90.81%</td>
      <td>-0.02%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>8120</td>
      <td>8118</td>
      <td>-2</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>820</td>
      <td>822</td>
      <td>+2</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30534/coverage_diff/stochastic_tools)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30534/coverage/stochastic_tools)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30534/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 32.06% is less than the suggested 90.0%
- `navier_stokes` new line coverage rate 86.81% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud committed:
add testing for combining Physics syntax with local syntax

lindsayad commented on framework/include/interfaces/BlockRestrictable.h:
```diff
@@ -127,6 +127,13 @@ class BlockRestrictable
 
+  /**
+   * Test if the supplied set of block names are valid for this object
+   * @param names A vector of SubdomainNames to check
```
```suggestion
   * @param names A set of SubdomainNames to check
```

lindsayad commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -735,10 +735,11 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+   * @param subdomain_names The names of the subdomains
    * @return The subdomain ids from the passed subdomain names.
    */
   std::vector<SubdomainID> getSubdomainIDs(const std::vector<SubdomainName> & subdomain_name) const;
+  std::set<SubdomainID> getSubdomainIDs(const std::set<SubdomainName> & subdomain_name) const;
```
```suggestion
  std::vector<SubdomainID> getSubdomainIDs(const std::vector<SubdomainName> & subdomain_names) const;
  std::set<SubdomainID> getSubdomainIDs(const std::set<SubdomainName> & subdomain_names) const;
```

GiudGiud replied:
auto-format add another line if we do it to the vector

lindsayad commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -168,13 +168,77 @@ class PhysicsBase : public Action, public InputParametersChecksUtils<PhysicsBase
+
+  // Routines to help with deciding when to create objects
+  /**
+   * Returns whether this Physics should create the variable or if it already exists
```
Which of these ors is it? Assuming just the first
```suggestion
   * \returns whether this Physics should create the variable. Will return \p false if the variable already exists
```

GiudGiud replied:
your wording. Let me add it

lindsayad commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -168,13 +168,77 @@ class PhysicsBase : public Action, public InputParametersChecksUtils<PhysicsBase
+                            const bool add_missing_subdomains);
+
+  /**
+   * Returns whether this Physics should create the IC or whether it already exists
```
same comment as above about ambiguity

lindsayad reviewed (commented):
Reviewed 6/33. Stopping for now. Coverage for the added code needs to increase by a lot. Currently at 32% in the framework unless the most recent round of pushes changes that. But I don't see any new tests

GiudGiud committed:
Address first review:
- lift ambiguity in docstring
- fix param names and description

GiudGiud committed:
Add all residual objects to theWarehouse
Add system name attributes to systems that were missing one

GiudGiud commented:
Added some tests. 
There is a decision to make wrt warehouses imo. I think we may want to decide to use theWarehouse for everything to standardize this code a little

lindsayad commented on framework/src/physics/PhysicsBase.C:
```diff
@@ -477,3 +495,241 @@ PhysicsBase::addPetscPairsToPetscOptions(
+bool
+PhysicsBase::isVariableFV(const VariableName & var_name) const
+{
+  const auto var = &_problem->getVariable(0, var_name);
+  if (const auto & field_var = dynamic_cast<MooseVariableFieldBase *>(var))
```
This already returns a `MooseVariableFieldBase`. No need for the dynamic cast

lindsayad commented on framework/src/physics/PhysicsBase.C:
```diff
@@ -477,3 +495,241 @@ PhysicsBase::addPetscPairsToPetscOptions(
+  const bool not_block_restricted =
+      (std::find(blocks.begin(), blocks.end(), "ANY_BLOCK_ID") != blocks.end()) ||
+      allMeshBlocks(blocks);
+  if (!var.blockRestricted() || (!not_block_restricted && var.hasBlocks(blocks)))
```
If `blocks` does indicate some kind of block restriction but `var.hasBlocks(blocks)` is `true`, it seems like we should still be returning `false` here

lindsayad commented on framework/src/physics/PhysicsBase.C:
```diff
@@ -477,3 +495,241 @@ PhysicsBase::addPetscPairsToPetscOptions(
+
+  // Deal with block restriction
+  bool has_all_blocks;
+  std::set<SubdomainName> blocks_covered;
```
Can you use IDs instead of names? This would make it unnecessary to do the set of IDs to vec of IDs to vec of names to set of names conversion. And also its easier/faster to compare integers than strings

GiudGiud replied:
ok let me try that!

lindsayad commented on framework/src/physics/PhysicsBase.C:
```diff
@@ -477,3 +495,241 @@ PhysicsBase::addPetscPairsToPetscOptions(
+  const auto time_vector_tag =
+      (sys_num < _problem->numNonlinearSystems())
+          ? var->sys().timeVectorTag()
+          // this is not quite correct. Many kernels can contribute to RHS time vector on paper
```
I don't understand why this makes what you're doing incorrect. Are there more than one time vector tags for linear systems? Sure multiple kernels can contribute to the time vector tag, but I still only think of there being one time vector tag

GiudGiud replied:
The problem is not the number of vector tags, it's whether all the kernels contributing to theRHS time are time derivative kernels. If they are not, then adding a time derivative could be fine

lindsayad commented on framework/src/physics/PhysicsBase.C:
```diff
@@ -477,3 +495,241 @@ PhysicsBase::addPetscPairsToPetscOptions(
+
+  // We just use the warehouse, it should cover every time derivative object type
+  bool all_blocks_covered = true, some_block_covered = false;
+  std::set<SubdomainName> blocks_covered;
```
same question about using IDs instead of names

lindsayad commented on framework/src/physics/PhysicsBase.C:
```diff
@@ -477,3 +495,241 @@ PhysicsBase::addPetscPairsToPetscOptions(
+          : dynamic_cast<LinearSystem *>(&var->sys())->rightHandSideTimeVectorTag();
+
+  // We just use the warehouse, it should cover every time derivative object type
+  bool all_blocks_covered = true, some_block_covered = false;
```
I've looked over three different methods with three different ways of doing this block intersection calculation. It sure would be nice if we could make this more uniform. I'm not saying it can be done for sure

GiudGiud replied:
uniformized now. There are two ways roughly now. Variables are a little simpler than ICs and time derivative block restriction wise

lindsayad commented on modules/navier_stokes/src/physics/WCNSFVFlowPhysics.C:
```diff
@@ -198,14 +198,26 @@ WCNSFVFlowPhysics::addSolverVariables()
-  if (variableExists(_pressure_name, true))
-    checkBlockRestrictionIdentical(_pressure_name,
-                                   getProblem().getVariable(0, _pressure_name).blocks());
+  const bool using_pinsfv_pressure_var =
```
do you mean the bernoulli variable?

GiudGiud replied:
changed

lindsayad commented on modules/navier_stokes/src/physics/WCNSFVFlowPhysics.C:
```diff
@@ -198,14 +198,26 @@ WCNSFVFlowPhysics::addSolverVariables()
+                                               "pressure_drop_form_factors"};
+      potentially_missed.insert(potentially_missed.end(), other_missed.begin(), other_missed.end());
+    }
+    reportPotentiallyMissedParameters(potentially_missed, "INSFVPressureVariable");
```
should you change the variable class name if `using_pinsfv_pressure_var`?

GiudGiud replied:
sure will be a tad clearer

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/ins/action/restart-block-restriction/tests:
```diff
@@ -9,16 +9,15 @@
-                 "NSFVAction. The restriction of the variable is: \(ANY_BLOCK_ID\) while the "
-                 "restriction for the action is: \(1, 2\)|Physics 'NavierStokesFV' and object "
-                 "'vel_x' have different block restrictions.)"
+    expect_err = "Block restriction of interpolator user object 'ins_rhie_chow_interpolator' \(1\) doesn't match the block restriction of variable 'vel_x'"
```
I guess this answers to my comment earlier in `PhysicsBase`; you're simply choosing not to support non-matching restrictions at this point, even if the variable is a superset

GiudGiud replied:
Yes I m not sure this is needed at this point.

lindsayad commented on test/tests/physics/block_restriction/tests:
```diff
@@ -0,0 +1,64 @@
+    input = 'diffusion_cg.i'
+    cli_args = "Variables/u/block='1 0' Physics/Diffusion/ContinuousGalerkin/diff/block='0 1'"
+    csvdiff = 'diffusion_cg_out_sample_0010.csv'
+    requirement = 'The system shall be able to define the variables outside the Physics shorthand syntax.'
```
```suggestion
    requirement = 'The system shall be able to define the variables outside shorthand physics syntax.'
```

lindsayad replied:
apply similarly throughout the file

GiudGiud commented on framework/src/physics/PhysicsBase.C:
```diff
@@ -477,3 +495,241 @@ PhysicsBase::addPetscPairsToPetscOptions(
+PhysicsBase::isVariableFV(const VariableName & var_name) const
+{
+  const auto var = &_problem->getVariable(0, var_name);
+  if (const auto & field_var = dynamic_cast<MooseVariableFieldBase *>(var))
+    return field_var->isFV();
+  return false;
```
```suggestion
  return field_var->isFV();
```

GiudGiud committed:
Address Alex review:
- use ids for checking block restriction easier
- use 'shorthand physics' instead of 'Physics shorthand'
- docstrings
- use subdomain Ids for APIs in the warehouses
- add some convenience PhyscicsBase APIs for block restriction printing/obtaining

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented on framework/include/warehouses/MooseObjectWarehouseBase.h:
```diff
@@ -90,6 +91,11 @@ class MooseObjectWarehouseBase
   bool hasObjects(THREAD_ID tid = 0) const;
   bool hasActiveObjects(THREAD_ID tid = 0) const;
+  bool hasObjectsForVariable(const VariableName & var_name, THREAD_ID tid /* = 0*/) const;
+  bool hasObjectsForVariableAndBlocks(const VariableName & var_name,
```
added

GiudGiud commented on framework/include/warehouses/MooseObjectWarehouseBase.h:
```diff
@@ -111,6 +117,10 @@ class MooseObjectWarehouseBase
   std::shared_ptr<T> getActiveObject(const std::string & name, THREAD_ID tid = 0) const;
   ///@}
 
+  /// Getter for objects that have the 'variable' set to a particular variable
```
added

GiudGiud commented on framework/include/warehouses/MooseObjectWarehouseBase.h:
```diff
@@ -457,6 +473,55 @@ MooseObjectWarehouseBase<T>::hasActiveObjects(THREAD_ID tid /* = 0*/) const
+      return true;
+    }
+  }
+  // No object has all the blocks, but one might overlap, which could be troublesome.
```
we want to define an IC on block 1 and 2.
There are no pre-existing IC on both blocks, but there is one on block 1. This is trouble

GiudGiud committed:
Add more comments in shouldCreateXYZ routine calls

lindsayad commented:
The coverage report seems way off

GiudGiud commented:
I'll rebase to see if it fixes it

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2941421/), step Verify coverage on 90e4e1b wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `combined` coverage rate (required: 37.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 84.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- `porous_flow` coverage rate 94.07% is less than the required 95.0%

GiudGiud commented:
Still wacky. 
I verified on a recipe we are running the same amounts of tests as devel for moose and 3 modules

GiudGiud commented:
Thanks for the review!
